extensions/amp-share-tracking/0.1/test/test-amp-share-tracking.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS-IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createIframePromise} from '../../../../testing/iframe';
+import {AmpShareTracking} from '../amp-share-tracking';
+import {Viewer} from '../../../../src/service/viewer-impl';
+import {Xhr} from '../../../../src/service/xhr-impl';
+import {toggleExperiment} from '../../../../src/experiments';
+import * as sinon from 'sinon';
+
+describe('amp-share-tracking', () => {
+  let sandbox;
+  let viewerForMock;
+  let xhrMock;
+
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    viewerForMock = sandbox.stub(Viewer.prototype, 'getFragment');
+    xhrMock = sandbox.stub(Xhr.prototype, 'fetchJson');
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  function getAmpShareTracking(optVendorUrl, optStubGetIncomingFragment,
+        optStubGetOutgoingFragment) {
+    return createIframePromise().then(iframe => {
+      toggleExperiment(iframe.win, 'amp-share-tracking', true);
+      const el = iframe.doc.createElement('amp-share-tracking');
+      if (optVendorUrl) {
+        el.setAttribute('data-href', optVendorUrl);
+      }
+      const ampShareTracking = new AmpShareTracking(el);
+      if (optStubGetIncomingFragment) {
+        sandbox.stub(ampShareTracking, 'getIncomingFragment_',
```

Comment:
Because testing the `getIncomingFragment_` and `getOutgoingFragment_` together would have many test cases. I separate the testing for only one method at a time to focus on one thing in a single test. I wrote 4 tests for `getIncomingFragment_`, 4 tests for `getOutgoingFragment_`. If I'm testing them together without stubbing, I would have 4*4=16 tests.